clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,331 @@
+		pingCheckers:  make(map[string]*pingChecker),
+	}
+}
+
+type serviceCreationLatencyMeasurement struct {
+	selector        *measurementutil.ObjectSelector
+	waitTimeout     time.Duration
+	stopCh          chan struct{}
+	isRunning       bool
+	selectorsString string
+	queue           workerqueue.Interface
+	client          clientset.Interface
+
+	lock          sync.Mutex
+	creationTimes *measurementutil.ObjectTransitionTimes
+	pingCheckers  map[string]*pingChecker
+}
+
+// Execute executes service startup latency measurement actions.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all-namespace scope is assumed.
+// "start" action starts observation of the services.
+// "waitForReady" waits until all services are reachable.
+// "gather" returns service created latency summary.
+func (s *serviceCreationLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultServiceCreationLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.start()
+	case "waitForReady":
+		return nil, s.waitForReady()
+	case "gather":
+		return s.gather()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (s *serviceCreationLatencyMeasurement) Dispose() {
+	if s.isRunning {
+		s.isRunning = false
+		close(s.stopCh)
+	}
+	s.queue.Stop()
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for _, pc := range s.pingCheckers {
+		pc.stop()
+	}
+	s.pingCheckers = make(map[string]*pingChecker)
+}
+
+// String returns a string representation of the metric.
+func (s *serviceCreationLatencyMeasurement) String() string {
+	return serviceCreationLatencyName + ": " + s.selector.String()
+}
+
+func (s *serviceCreationLatencyMeasurement) start() error {
+	if s.isRunning {
+		klog.Infof("%s: wait for controlled pods measurement already running", s)
+		return nil
+	}
+	klog.Infof("%s: starting wait for controlled pods measurement...", s)
+
+	s.isRunning = true
+	s.stopCh = make(chan struct{})
+
+	i := informer.NewInformer(
+		s.client,
+		"services",
+		s.selector,
+		func(oldObj, newObj interface{}) {
+			f := func() {
+				s.handleObject(oldObj, newObj)
+			}
+			s.queue.Add(&f)
+		},
+	)
+	return informer.StartAndSync(i, s.stopCh, informerSyncTimeout)
+}
+
+func (s *serviceCreationLatencyMeasurement) waitForReady() error {
+	for {
+		created, ipAssigned, reachable := 0, 0, 0
+		select {
+		case <-time.After(s.waitTimeout):
+			klog.Errorf("%s: waiting for services timeout")
+			return fmt.Errorf("waiting for services timeout")
```

Comment:
Don't log error if you return it - if you return an error it's responsibility of a called to handle it (whether they want to log it or not).

Reply:
Done.